ibis/expr/builders.py

```diff
@@ -176,6 +176,8 @@ def _maybe_cast_boundary(self, boundary, dtype):
 
     def _maybe_cast_boundaries(self, start, end):
         if start and end:
+            if dt.same_base_type(start.dtype, end.dtype):
+                return start, end
             dtype = dt.higher_precedence(start.dtype, end.dtype)
             start = self._maybe_cast_boundary(start, dtype)
             end = self._maybe_cast_boundary(end, dtype)
```

Comment:
Just curious, does removing these lines not address the problem?

Reply:
I tried doing that, and that breaks the windowing logic for at least some backends. For example, if the window is something like `-ibis.interval(seconds=20), 0` where one of the boundary is an interval and the other one is an int (most commonly 0), it ends up erroring out when executed on the backend (e.g. pandas). So essentially, instead of baking the casting logic into the IR we would move the casting logic out to backends where casting is necessary, which is doable but seems like a more complex solution

Reply:
I dislike introducing `is_same_base_type()` api just to overcome this issue. There are basically two kind of inputs are supported here, numeric and interval. Could we spell out all combinations instead, something like:

```py
def _maybe_cast_boundaries(self, start, end):
    if start.dtype.is_interval() and end.dtype.is_numeric():
        return start, ops.Cast(end, start.dtype)
    elif start.dtype.is_numeric() and end.dtype.is_interval():
        return ops.Cast(start, end.dtype), end
    else:
        return start, end
```

Reply:
Sure, I think that's fair.

> Also just check for `rlz.comparable(start, end)` during the `WindowFrame` validation.
I think this would yield the same result as above, when we were to remove the `_maybe_cast_boundaries()` logic entirely. At this point it should be required that window start and window end are both numeric or both intervals.

Reply:
Indeed, then we should validate for that.